### PR TITLE
samples: nrf9160: modem_shell: link cmd updates

### PR DESCRIPTION
**Pull request by jhirsi**
Following updates and fixes to link command:
- getopt usage fix
- link connect authentication support
- PDP contexts are reactivated when needed
- registering always to RSRP when going to normal mode
- link ncellmeas continuous mode.

### CODE REVIEW

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 1]
```diff
@@ -1280,9 +1301,39 @@ int link_shell(const struct shell *shell, size_t argc, char **argv)
 			link_modem_tau_notifications_unsubscribe();
 		}
 		break;
-	case LINK_CMD_CONNECT:
-		ret = link_shell_pdn_connect(shell, apn, family);
-		break;
+	case LINK_CMD_CONNECT: {
+		struct link_shell_pdn_auth auth_params;
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/link/link_shell.c:1305]: This is a bit of a common problem in this file that there are quite long code blocks within this switch case. I'd like to have this type of code handled in separate functions.

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 2]
```diff
@@ -809,7 +823,8 @@ int link_shell(const struct shell *shell, size_t argc, char **argv)
 				shell_error(
 					shell,
 					"Not a valid number for --threshold_time (milliseconds).");
-				return -EINVAL;
+				ret = -EINVAL;
+				goto return_without_usage;
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/link/link_shell.c:827]: Is this change clearer than the original code, or what's the intention here? I'd say it's easier to read if we just return here, or am I missing some functional change?

File: samples/nrf9160/modem_shell/src/link/link_shell_pdn.h [hunk 3]
```diff
@@ -10,13 +10,24 @@
...
+	enum pdn_auth method;
+	char *user;
+	char *password;
+};
+
 void link_shell_pdn_init(const struct shell *shell);
 
 int link_shell_pdn_connect(const struct shell *shell, const char *apn_name,
-			    const char *family_str);
+			   const char *family_str,
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/link/link_shell_pdn.h:22]: Indentation should match with the previous line based on what's seen in nrf code.

File: samples/nrf9160/modem_shell/src/link/link_shell_pdn.c [hunk 4]
```diff
@@ -169,13 +177,78 @@ const char *link_pdn_lib_family_to_string(enum pdn_fam pdn_family,
...
+	struct link_shell_pdn_info *iterator = NULL;
+
+	/* Check if already in list, if there; then return an existing one: */
+	SYS_DLIST_FOR_EACH_CONTAINER(&pdn_info_list, iterator, dnode) {
+		if (iterator->cid == cid) {
+			return iterator;
+		}
+	}
+
+	/* Not already at list, let's add it: */
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/link/link_shell_pdn.c:192]: Not sure we need colon in the end of each comment. I wouldn't use anything there.

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 5]
```diff
@@ -231,10 +230,11 @@ static const char link_rsrp_usage_str[] =
...
 
 static const char link_ncellmeas_usage_str[] =
-	"Usage: link ncellmeas --single | --cancel\n"
+	"Usage: link ncellmeas --single | --continuous |--cancel\n"
 	"Options:\n"
-	"      --single, Start a neighbor cell measurement and report result\n"
-	"      --cancel, Cancel/Stop neighbor cell measurement if still on going\n";
+	"      --single,     Start a neighbor cell measurement and report result\n"
+	"      --continuous, Start continuous neighbor cell measurement mode and report result\n"
+	"      --cancel,     Cancel/Stop neighbor cell measurement if still ongoing\n";
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/link/link_shell.c:237]: I know this is older stuff but seeing that stop is used in the entire code, should we rename `--cancel` to `--stop`?

File: samples/nrf9160/modem_shell/src/link/link.c [hunk 6]
```diff
@@ -34,32 +34,111 @@
...
 /******************************************************************************/
 
-static void link_modem_info_work(struct k_work *unused)
+static void link_api_activate_mosh_contexts(
+	struct pdn_activation_status_info pdn_act_status_arr[], int size)
+{
+	int i, esm;
+
+	/* Check that all context created by mosh link connect are active and if not,
+	 * then activate:
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/link/link.c:78]: There are many of these lines that would fit into 100 character lines. I'll leave it to you whether to change them or not but think about it especially in future changes.

File: samples/nrf9160/modem_shell/src/link/link_shell_pdn.c [hunk 7]
```diff
@@ -202,17 +274,25 @@ int link_shell_pdn_connect(const struct shell *shell, const char *apn_name,
...
 	}
 
+	pdn_id = pdn_id_get(cid);
+
+	/* Add to PDN list: */
+	new_pdn_info = link_shell_pdn_info_list_add(pdn_id, cid);
+	if (new_pdn_info == NULL) {
+		shell_error(
+			shell,
+			"ltelc_pdn_init_and_connect: could not add new PDN socket to list\n");
```
Comment (rlubos) [inline samples/nrf9160/modem_shell/src/link/link_shell_pdn.c:302]: Is `ltelc_pdn_init_and_connect` correct in this context? It doesn't seem to me, unless I miss something.

File: samples/nrf9160/modem_shell/src/link/link_shell_pdn.h (hunk 3 above)
Comment (rlubos) [reply samples/nrf9160/modem_shell/src/link/link_shell_pdn.h:22]: It does in this particular case, it's github diff preview that messes with tab indentation.

Comment (trantanen) [reply samples/nrf9160/modem_shell/src/link/link_shell_pdn.h:22]: ok

File: samples/nrf9160/modem_shell/src/link/link_shell.c (hunk 2 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell.c:827]: Removed the change.

File: samples/nrf9160/modem_shell/src/link/link_shell_pdn.c (hunk 4 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell_pdn.c:192]: Removed those

File: samples/nrf9160/modem_shell/src/link/link_shell_pdn.c (hunk 7 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell_pdn.c:302]: typo, fixed

File: samples/nrf9160/modem_shell/src/link/link.c (hunk 6 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link.c:78]: Yes, did some fixes elsewhere but this particular one does not fit to 100.

File: samples/nrf9160/modem_shell/src/link/link_shell.c (hunk 5 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell.c:237]: stop was one of the options, however this was discussed and the decision earlier was that this to be cancel. However, fixed help test from main level to be cancel.

File: samples/nrf9160/modem_shell/src/link/link_shell.c (hunk 1 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell.c:1295]: Yes, agree, a common problem in this file. Parsing is done mainly here, and not sure does splitting in this case makes this better. Left as it is for now.

**Review by trantanen — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

File: samples/nrf9160/modem_shell/src/link/link.c [hunk 8]
```diff
@@ -34,32 +34,111 @@
...
+struct pdn_activation_status_info {
+	bool activated;
+	uint8_t cid;
+};
+
 /* System work queue for getting the modem info that ain't in lte connection ind: */
-static struct k_work modem_info_work;
+static struct k_work registered_work;
 
 /* Work queue for signal info: */
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/link/link.c:54]: Nit, `k_work` is not a work queue, but a work (item).

**Review by tokangas — APPROVED** [review]
Looks good, only a couple of nits.

File: samples/nrf9160/modem_shell/src/link/link.c [hunk 9]
```diff
@@ -53,9 +53,22 @@ static struct k_work modem_info_work;
 static struct k_work modem_info_signal_work;
 #define LINK_RSRP_VALUE_NOT_KNOWN -999
 static int32_t modem_rsrp = LINK_RSRP_VALUE_NOT_KNOWN;
+#endif
+
+/* Work queue for continuous neighbor cell measurements: */
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link.c]: Fixed.

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 10]
```diff
@@ -130,10 +130,14 @@ static const char link_defcontauth_usage_str[] =
 	"                      2 (CHAP)\n";
 
 static const char link_connect_usage_str[] =
-	"Usage: link connect --apn <apn str> [--family <pdn family str>]\n"
+	"Usage: link connect --apn <apn str> [--family <pdn family str>][auth options]\n"
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell.c]: Fixed.

[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell.c]: Fixed.

File: samples/nrf9160/modem_shell/src/link/link.c [hunk 11]
```diff
@@ -34,26 +34,27 @@
...
 
 extern bool uart_disable_during_sleep_requested;
 
-#if defined(CONFIG_MODEM_INFO)
+struct pdn_activation_status_info {
+	bool activated;
+	uint8_t cid;
+};
+
 /* System work queue for getting the modem info that ain't in lte connection ind: */
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link.c]: Fixed.

**Pull request merged**